lib/checkunusedvar.cpp

```diff
@@ -1604,15 +1604,47 @@ bool CheckUnusedVar::isFunctionWithoutSideEffects(const Function& func, const To
         return false;
     }
 
-    for (const Token* argsToken = functionUsageToken->next(); !Token::simpleMatch(argsToken, ")"); argsToken = argsToken->next()) {
-        const Variable* argVar = argsToken->variable();
-        if (argVar && argVar->isGlobal()) {
-            return false; // TODO: analyze global variable usage
+    // search for global variables in arguments list
+    std::vector<unsigned int> globalVarsIndexes;
+    unsigned int usageVarIndex = 0;
+    const Token* argsClosingBracketToken = functionUsageToken->next()->link();
```

Comment:
I don't understand this globalVarsIndexes stuff..

imagine that x,y,z are global variables. and then we have this function call: `dostuff(1+x+y, z);`. is it intentional that x will get usage index 1, z will get usage index 3?

Spontanously I would first collect the function arguments with getArguments. then you will get 2 argument expressions `1+x+z` and `z`.. and then I would try to determine if each expression is "global" and check the function body how that corresponding parameter is used..

Reply:
> I don't understand this globalVarsIndexes stuff..

The idea was to parse arguments at function usage and found which of them are global variables. Save this data as argument indexes. Then treat arguments at these indexes as global variables in function body.
At least I didn't think up any better way than this one.

> imagine that x,y,z are global variables. and then we have this function call: `dostuff(1+x+y, z);`. is it intentional that x will get usage index 1, z will get usage index 3?

No, this is my fault. As you can see, I don't have such test case. `1+x+y` should have usage index 0, `z` should have index 1.
I'll will fix this and will return with a new PR revision.

> Spontanously I would first collect the function arguments with getArguments. then you will get 2 argument expressions `1+x+z` and `z`.. and then I would try to determine if each expression is "global" and check the function body how that corresponding parameter is used..

Thanks, I will try this. You know, some solutions are hard to be found without proper API knowledge. =)

Reply:
> Spontanously I would first collect the function arguments with getArguments.

As I see, `Function::getArgumentVar()` (the closest in meaning method I found) returns definition arguments variables.


```
void func(int arg1, int arg2);
...
func(var1, var2);
```
`arg1` and `arg2` vars will be returned, not `var1` and `var2`.

Reply:
Hello!
Unfortunately I will not be able to finish it now. I will return to this PR after a few weeks.

Reply:
Thanks for the information.

Reply:
>In astutils.cpp :
std::vector<const Token *> getArguments(const Token *ftok)

Seems like this function also doesn't handle expressions in arguments (AST parent only handled?).
Code line from test to be analyzed: `            "   C() : x(func(y+z+2, buf, 3)) {}\n"`
Function usage:
```
    std::cout << "functionUsageToken: " << functionUsageToken->str() << std::endl;
    std::vector<const Token*> args = getArguments(functionUsageToken);
    std::cout << "args: " << std::endl;
    for (const Token* arg : args) {
        std::cout << arg->str() << std::endl;
    }
```
Console output:
```
functionUsageToken: func
args: 
+
buf
3
```

I had some clumsy arguments walkthrough implementation, which handled issue you mentioned in first comment under this thread. I will return back to stabilizing it, if you don't have any other already implemented solution that may fit here.

Reply:
I pushed one more commit with this issue fix. Code still need some refactoring and more tests need to be added. You can review it as a PoC now.

Reply:
Hm, I noticed that you used `arg->expressionString()` instead of plain `arg->str()` in your example. I will recheck.

Reply:
Unfortunately, I didn't manage to make `isVariableChangedByFunctionCall()` work as expected for my tests. Maybe it should be updated for some specific situations. Anyway, not in this PR.